amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -0,0 +1,216 @@
+    connectionTimeout.foreach(factory.setConnectionTimeout)
+    handshakeTimeout.foreach(factory.setHandshakeTimeout)
+    shutdownTimeout.foreach(factory.setShutdownTimeout)
+    networkRecoveryInterval.foreach(factory.setNetworkRecoveryInterval)
+    automaticRecoveryEnabled.foreach(factory.setAutomaticRecoveryEnabled)
+    topologyRecoveryEnabled.foreach(factory.setTopologyRecoveryEnabled)
+    exceptionHandler.foreach(factory.setExceptionHandler)
+
+    factory.newConnection(hostAndPortList.map(hp => new Address(hp._1, hp._2)).asJava)
+  }
+
+}
+
+object AmqpDetailsConnectionProvider {
+
+  def apply(host: String, port: Int): AmqpDetailsConnectionProvider =
+    AmqpDetailsConnectionProvider(List((host, port)))
+
+  /**
+   * Java API
+   */
+  def create(host: String, port: Int): AmqpDetailsConnectionProvider =
+    AmqpDetailsConnectionProvider(host, port)
+}
+
+object AmqpCredentials {
+
+  /**
+   * Java API
+   */
+  def create(username: String, password: String): AmqpCredentials =
+    AmqpCredentials(username, password)
+}
+
+final case class AmqpCredentials(username: String, password: String) {
+  override def toString = s"Credentials($username, ********)"
+}
+
+object AmqpCachedConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(provider: AmqpConnectionProvider): AmqpCachedConnectionProvider =
+    AmqpCachedConnectionProvider(provider: AmqpConnectionProvider)
+
+  /**
+   * Java API
+   */
+  def create(provider: AmqpConnectionProvider, automaticRelease: Boolean): AmqpCachedConnectionProvider =
+    AmqpCachedConnectionProvider(provider: AmqpConnectionProvider, automaticRelease)
+
+  private sealed trait State
+  private case object Empty extends State
+  private case object Connecting extends State
+  private final case class Connected(connection: Connection, clients: Int) extends State
+  private case object Closing extends State
+}
+
+final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider, automaticRelease: Boolean = true)
+    extends AmqpConnectionProvider {
+
+  import akka.stream.alpakka.amqp.AmqpCachedConnectionProvider._
+  private val state = new AtomicReference[State](Empty)
+
+  @tailrec
+  override def get: Connection = state.get match {
+    case Empty =>
+      if (state.compareAndSet(Empty, Connecting)) {
+        val connection = provider.get
+        state.compareAndSet(Connecting, Connected(connection, 1))
+        connection
+      } else get
+    case c @ Connected(connection, clients) =>
+      if (state.compareAndSet(c, Connected(connection, clients + 1))) connection
+      else get
+    case _ => get
+  }
+
+  @tailrec
+  override def release(connection: Connection): Unit = state.get match {
+    case c @ Connected(cachedConnection, clients) =>
+      if (cachedConnection != connection)
+        throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
+
+      if (state.compareAndSet(c, Closing)) {
```

Comment:
It should not always switch to `Closing`, only if it is the last client and automatic release is enabled it should try to win the cas for `Closing`, in the other case it should just update state.compareAndSet(c, c.copy(clients = c.clients - 1))` and it must check if that succeeded and recurse if it failed (because another client may have `released` or called `get` concurrently)